### PR TITLE
feat: Upgrade to Pydantic V2

### PR DESCRIPTION
**Pull request by michael-genson**
## What type of PR is this?

_(REQUIRED)_

- bug
- feature

## What this PR does / why we need it:

_(REQUIRED)_

This PR is a monster. Pydantic provides a tool to make _some_ model changes (`bump-pydantic`), but it doesn't cover most things. Thankfully, their migration guide covers just about everything, save for some of the grosser stuff: https://docs.pydantic.dev/latest/migration/

Most changes are pretty straightforward, but there are a few... "special" changes that are a bit hacky:
- We use a custom in-house controller for our FastAPI routes. Part of the magic-controller-stuff includes walking params and handling them in various ways, including certain handling for classvars. We relied on Pydantic's `is_classvar` for this, but that doesn't exist anymore. Rather than attempt to figure out how that part of the codebase works, I just copied it over from Pydantic V1
- Similarly, when parsing datetime strings, we relied on Pydantic's `datetime_parse` module/method to handle this. I tried using a recommended substitution, but got different results than we were expecting (timezones were getting mishmashed somewhere and tests were failing) so I, again, just copied Pydantic V1's implementation
- The `GetterDict` isn't a thing anymore, so all of our custom getter-dicts had to be replaced. I basically replaced them with field validators (Pydantic V2's replacement for `@validator`), and fixed some other edgecases (particularly with `Group`/`User.group` being the group name)

Also, Somehow the format for `PostgresDsn` changed. I had to tweak it to get postgres to work right, but if you could take a look and make sure I didn't screw that up, that'd be great.

## Which issue(s) this PR fixes:

_(REQUIRED)_

N/A, will close some renovate bot PRs

### CODE REVIEW

File: mealie/core/settings/db_providers.py [hunk 1]
```diff
@@ -38,15 +39,19 @@ class PostgresProvider(AbstractDBProvider, BaseSettings):
     POSTGRES_PORT: str = "5432"
     POSTGRES_DB: str = "mealie"
 
+    model_config = SettingsConfigDict(arbitrary_types_allowed=True, extra="allow")
+
     @property
     def db_url(self) -> str:
         host = f"{self.POSTGRES_SERVER}:{self.POSTGRES_PORT}"
-        return PostgresDsn.build(
```
Comment (michael-genson) [inline mealie/core/settings/db_providers.py:44]: PostgresDsn changes, pretty sure this is right

File: mealie/routes/_base/controller.py [hunk 2]
```diff
@@ -47,6 +46,25 @@ def _cbv(router: APIRouter, cls: type[T], *urls: str, instance: Any | None = Non
     return cls
 
 
+# copied from Pydantic V1 Source: https://github.com/pydantic/pydantic/blob/1c91c8627b541b22354b9ed56b9ef1bb21ac6fbd/pydantic/v1/typing.py
```
Comment (michael-genson) [inline mealie/routes/_base/controller.py:49]: Gross copy from Pydantic V1

File: mealie/schema/_mealie/datetime_parse.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
+"""
```
Comment (michael-genson) [inline mealie/schema/_mealie/datetime_parse.py:1]: Copied from Pydantic V1. I wasn't sure where to put it so I put it in `schema/_mealie`

**Comment by michael-genson** [comment]
I tried to highlight some of the bigger changes in the comments, but this PR is pretty massive

**Review by hay-kot — DISMISSED** [review]
LGTM. This was such a huge lift. Thanks for taking this on. 

I'll give you $50 if you didn't break anything 😉

**Comment by michael-genson** [comment]
> if you didn't break anything

Optimistic

**Review by hay-kot — APPROVED** [review]

**Pull request merged**